java-checks/src/main/java/org/sonar/java/checks/StaticMethodCheck.java

```diff
@@ -106,28 +105,134 @@ private static boolean isExcluded(MethodTree tree) {
 
-  private boolean fromInstance(Symbol owner) {
-    for (Symbol outerClass : outerClasses) {
-      Type ownerType = owner.type();
-      if (ownerType != null) {
-        if (owner.equals(outerClass) || outerClass.type().isSubtypeOf(ownerType)) {
+  private static class MethodReference {
+
+    private final Symbol.MethodSymbol methodSymbol;
+    private final Symbol methodScopeOwner;
+    private boolean nonStaticReference = false;
+
+    MethodReference(Symbol.MethodSymbol symbol) {
+      methodSymbol = symbol;
+      methodScopeOwner = scopeOwner(methodSymbol);
+      if (methodScopeOwner != null && methodScopeOwner.isTypeSymbol()) {
+        nonStaticReference = !methodScopeOwner.isStatic() && !methodScopeOwner.owner().isPackageSymbol();
+      }
+    }
+
+    @CheckForNull
+    private static Symbol scopeOwner(Symbol symbol) {
+      Symbol owner = symbol.owner();
+      while (owner != null) {
+        if (owner.isPackageSymbol() || owner.isTypeSymbol() || owner.isMethodSymbol()) {
+          break;
+        }
+        owner = owner.owner();
+      }
+      return owner;
+    }
+
+    @CheckForNull
+    private static Symbol getPackage(Symbol symbol) {
+      Symbol owner = symbol.owner();
+      while (owner != null) {
+        if (owner.isPackageSymbol()) {
+          break;
+        }
+        owner = owner.owner();
+      }
+      return owner;
+    }
+
+    void setNonStaticReference() {
+      nonStaticReference = true;
+    }
+
+    boolean hasNonStaticReference() {
+      return nonStaticReference;
+    }
+
+    void checkSymbol(Symbol symbol) {
+      if (nonStaticReference || symbol == methodSymbol || symbol.isStatic()) {
+        return;
+      }
+      final Symbol scopeOwner = scopeOwner(symbol);
+      if (isConstructor(symbol)) {
+        checkConstructor(scopeOwner);
+      } else if (scopeOwner != null) {
+        checkNonConstructor(symbol, scopeOwner);
+      }
+    }
+
+    public void checkConstructor(final Symbol constructorClass) {
+      if (!constructorClass.isStatic()) {
+        Symbol methodPackage = getPackage(methodScopeOwner);
+        Symbol constructorPackage = getPackage(constructorClass);
+        if (Objects.equals(methodPackage, constructorPackage) && !constructorClass.owner().isPackageSymbol()) {
+          setNonStaticReference();
+        }
+      }
+    }
+
+    public void checkNonConstructor(final Symbol symbol, final Symbol scopeOwner) {
+      if (scopeOwner.isMethodSymbol()) {
+        return;
+      }
+      if (hasLocalAccess(methodScopeOwner, scopeOwner)) {
+        setNonStaticReference();
+      }
+    }
+
+    private boolean isConstructor(Symbol symbol) {
+      return "<init>".equals(symbol.name());
+    }
+
+    private static boolean hasLocalAccess(Symbol scope, Symbol symbol) {
+      if (scope == symbol) {
+        return true;
+      }
+      if (scope.isTypeSymbol() && symbol.isTypeSymbol()) {
+        // FIXME use erasure until Parameterized types are correctly supported!
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Take the required action to fix the issue indicated by this comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1134)